### ISSUE TITLE
Flakes: per-system configuration

### ISSUE THREAD

**Issue by wmertens** [nix-community/home-manager#3075]
With flakes, the system has to be embedded in the pkgs selection. Instead, it would be easier to reuse configurations across platforms by having a setup like apps and packages, where the second level is the system.

Then `pkgs` can be optional.

So a config would go from

```nix
    {
      homeConfigurations = {
        ${username} = home-manager.lib.homeManagerConfiguration rec {
          pkgs = nixpkgs.legacyPackages.${system};
          modules = [
            # ...
          ];
        };
      };
    };
```

to

```nix
    flake-utils.lib.eachDefaultSystem (system: {
      homeConfigurations = {
        ${username} = home-manager.lib.homeManagerConfiguration rec {
          modules = [
            # ...
          ];
        };
      };
    })
```

I suppose both configurations could be supported by checking if the current system is an attribute in the home-manager script

**Comment by sepiabrown** [comment]
I also want this feature!

**Comment by bobvanderlinden** [comment]
I also ran into this inconvenience. This part of my config became a bit strange due to this restriction: https://github.com/bobvanderlinden/nixos-config/blob/bdfd8d94def9dc36166ef5725589bf3d7ae2d233/flake.nix#L38-L46

It now also isn't possible to share this home-manager configuration with OSX.

**Comment by ambroisie** [comment]
Still relevant.

**Comment by scottwillmoore** [comment]
Okay, so I'm new to Nix, but I may have a solution.

When I ran `home-manager switch --flake .#scott`, with no exported `homeConfigurations.scott`, I got this output...

`error: flake 'git+file:///home/scott/dotfiles' does not provide attribute 'packages.x86_64-linux.homeConfigurations."scott".activationPackage', 'legacyPackages.x86_64-linux.homeConfigurations."scott".activationPackage' or 'homeConfigurations."scott".activationPackage'`.

So, I tried to export `homeConfigurations.scott` as `packages.${system}.homeConfigurations.scott` instead, and this **does appear to work**. I don't think this behaviour comes from home manager, but instead from the Nix CLI, when `nix build` is invoked by home manager which is somewhat [documented](https://nixos.wiki/wiki/Flakes#Output_schema).

The work around explained by @nixos-discourse will also work, although a little uglier.

It still might be an idea to consider support for `homeConfigurations.${system}.scott`, but this is up to the home manager contributors.

**Comment by roberth** [comment]
The current interface seems fairly consistent with the attributes for NixOS, the idea being that a `nixosConfigurations.<name>` is as _specific_ as it gets; typically a specific host. `nixosModules.<name>` on the other hand can be reused in any configuration that needs it, including different cpu architectures, etc.

Perhaps for your use case, an attribute like `nixosModules.<name>` makes more sense, considering that you seem to want to use it in more than one configuration? This way the flake schema remains consistent with that of NixOS.

**Comment by edrex** [comment]
I think homeConfigurations is more like packages than nixosConfigurations. The reason each nixosConfiguration specifies system is that it is meant to run on specific hardware, whereas the point of hm is to make your home profile portable across hosts.

**Comment by edrex** [comment]
Unfortunately, the workaround detailed by @scottwillmoore doesn't work in a `flake-parts` context, since [the option](https://github.com/hercules-ci/flake-parts/blob/main/modules/packages.nix) expects values of type package:

```
error: A definition for option `perSystem.x86_64-linux.packages.homeConfigurations' is not of type `package'.
```

**Comment by Shados** [comment]
> The reason each nixosConfiguration specifies system is that it is meant to run on specific hardware, whereas the point of hm is to make your home profile portable across hosts.

`homeConfigurations` *may* also be specialised to specific hardware. For example, the HM config for my desktop has a number of differences compared to the config for my laptop, due to the difference in the number, size, resolution, and viewing distance of available monitors. The set of packages installed also varies between them, because they're used for different purposes.

Sharing configuration across `homeConfigurations`, and across a variety of `system`s, is already pretty trivial in a flake, for example I do something along the lines of:
```nix
{
  description = "Home Manager configuration of shados";

  inputs = {
    nixpkgs.url = "nixpkgs";
    home-manager = {
      url = github:nix-community/home-manager;
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };

  outputs = { nixpkgs, home-manager, ... } @ inputs:
    let
      mkHomeConfig = machineModule: system: home-manager.lib.homeManagerConfiguration {
        pkgs = import nixpkgs {
          inherit system;
        };

        modules = [
          ./sharedConfig
          machineModule
        ];

        extraSpecialArgs = {
          inherit inputs system;
        };
      };
    in {
      homeConfigurations."shados@desktop" = mkHomeConfig ./machine/desktop.nix "x86_64-linux";
      homeConfigurations."shados@laptop" = mkHomeConfig ./machine/laptop.nix "x86_64-linux";
      homeConfigurations."shados@work" = mkHomeConfig ./machine/work.nix "x86_64-darwin";
      homeConfigurations."shados@arm-vm" = mkHomeConfig ./machine/arm-vm.nix "aarch64-linux";
    };
}
```

Anything truly portable lives in the `./sharedConfig` module (or its imported modules), and per-device config lives in files under `./machine`.

If your HM config is intended to be 100% portable and identical across machines, you could get something pretty close to what @wmertens originally wanted:
```nix
      homeConfigurations = with nixpkgs.lib; listToAttrs (flip map flake-utils.lib.defaultSystems
        (system: nameValuePair "${username}-${system}" home-manager.lib.homeManagerConfiguration {
          pkgs = import nixpkgs { inherit system; };
          modules = [
            # ...
          ];
        }));
```

And then build with `home-manager switch --flake ~/.config/home-manager#username-x86_64-linux`, and drop the `--flake` option on subsequent calls, as per usual.

-----

If anything, I would say a better change would be for `home-manager.lib.homeManagerConfiguration` to explicitly take a `system` argument and a bare `nixpkgs` argument, instead of an instantiated `pkgs`. This would bring HM more in line with nixpkgs, and I think this would also make it easier to fix #2942 & #2954...

**Comment by martin-braun** [comment]
Oh this was a pitfall. I was wondering why it was complaining 

> flake ... does not provide attribute 'packages.x86_64-linux.homeConfigurations."<username>".activationPackage', 'legacyPackages.x86_64-linux.homeConfigurations."<username>".activationPackage' or 'homeConfigurations."<username>".activationPackage'

because I didn't understand that the (standalone) `home-manager` acts like `nixos-rebuild`, but instead of picking the `hostName` from the `nixosConfiguration`, it picks the `username` from `homeConfigurations`. 

So I have to `home-manager switch --flake .#<username>-x86_64-linux` to support multiple platforms, ahhh.

I don't understand why the `system` can't be read and used prior the flake resolve. That would make things easier, on system flakes as well.

**Comment by x10an14** [comment]
> (...)
> because I didn't understand that the (standalone) `home-manager` acts like `nixos-rebuild`, but instead of picking the `hostName` from the `nixosConfigurations`, it picks the `username` from `homeConfigurations`.
> 
> So I have to `home-manager switch --flake .#<username>-x86_64-linux` to support multiple platforms, ahhh.
> 
> I don't understand why the `system` can't be read and used prior the flake resolve. That would make things easier, on system flakes as well.

`home-manager switch` (without any other arguments) works for me when I have `<path.to.flake>#homeConfigurations."$USER@$(hostname)".activationPackage`? 

Maybe because it (my HM nix setup/repo) always imports all HM modules for all systems (read: any configured Linux user/combo) through a `home.nix` in root of repo, but I didn't think it has to be that way had.

A symlink is needed though to `~/.config/home-manager`, pointing to config repo I'm pretty sure.

**Comment by martin-braun** [comment]
@x10an14 I couldn't figure out how to do that. For me, it will always try to find the configuration of the `username` and only the `username`. I've seen a lot of configs that follow the `$USER@$(hostname)` scheme, but I couldn't pin point the key to toggle this behavior, unless passing `'.#'$USER@${hostname}`.

My situation is probably different and that might be relevant, because my configuration is strictly separated. My HM profile is a single-user multi-host configuration that I want to resolve by hostname only, similarly how Nix is doing it for the system flakes in `/etc/nixos/` on `sudo nixos-rebuild switch`.

I think my recent issue #4246 is a dub, so I will quote myself: 

> It would be nice if I could tell the `home-server` in my configuration to look for the hostname instead of username in `homeConfigurations`. I know I can use the `--flake` attribute with `#<hostname>` to get what I want, but it adds necessary tooling that I feel could be solved within the configuration itself.
> 
> Why would I want that? Since we can store user configurations in `$HOME`, it's reasonable to maintain single-user repositories (one user per repo) for multiple hosts, instead of the common case to have a per-user (multi user) repo for many hosts, which often puts everything together (incl. system configuration).
> 
> Imho, user and system should be separated, even user and another user (different persons) should be separated. So for me personally, it would make sense if `home-manager switch` looks for `homeConfiguration."<hostName>"` instead of `homeConfiguration."<username>"`.
> 
> Or why not just looking for `homeConfiguration."<username>@<hostName>"`, then `homeConfiguration."<username>"`, then `homeConfiguration."<hostName>"`, in that order to give much flexibility. As soon as a configuration is found, it will be used.

Lastly, I want to emphasize that I don't like any work arounds that require more abstraction. I'm happy to learn Flakes in their purest form, I don't want to involve additional abstractions, I even keep myself from defining any functions to re-use (i.e. `mkHomeConfiguration`) to keep the declaration flow as simple as possible.

For now, I will just have to pass more arguments to `home-server`, but hopefully my opinion will matter in the slightest. Perhaps one will acknowledge it as valuable feedback of new-comers and their problems adapting Nix in general, but I'm still too inexperienced to give a fundamental opinion to be honest.

Thank you.

**Comment by ambroisie** [comment]
Still relevant.

**Comment by jzbor** [comment]
Still relevant.

**Comment by zeorin** [comment]
Still relevant.

**Comment by soulomoon** [comment]
Still relevant 🥲

**Comment by FBIGlowie** [comment]
still unrelevant

**Comment by lz37** [comment]
> Okay, so I'm new to Nix, but I may have a solution.
> 
> When I ran `home-manager switch --flake .#scott`, with no exported `homeConfigurations.scott`, I got this output...
> 
> `error: flake 'git+file:///home/scott/dotfiles' does not provide attribute 'packages.x86_64-linux.homeConfigurations."scott".activationPackage', 'legacyPackages.x86_64-linux.homeConfigurations."scott".activationPackage' or 'homeConfigurations."scott".activationPackage'`.
> 
> So, I tried to export `homeConfigurations.scott` as `packages.${system}.homeConfigurations.scott` instead, and this **does appear to work**. I don't think this behaviour comes from home manager, but instead from the Nix CLI, when `nix build` is invoked by home manager which is somewhat [documented](https://nixos.wiki/wiki/Flakes#Output_schema).
> 
> The work around explained by @nixos-discourse will also work, with `homeConfigurations."${system}-scott`, although a little uglier.
> 
> It still might be an idea to consider support for `homeConfigurations.${system}.scott`, but this is up to the home manager contributors.
> 
> TL;DR: It appears that `packages.${system}.homeConfigurations.scott` will work.

genus!!! it works

**Comment by davidmh** [comment]
Still relevant.

**Comment by pabroux** [comment]
For anyone wanting a system agnostic setup for Home Manager and not wanting to give extra argument to `home-manager switch`, here what should look like your `flake.nix`:

``` nix
{
  description = "unvX: a simple Home Manager configuration";

  inputs = {
    # nixpkgs, home-manager and flake-utils are by default loaded since they are
    # available in the global flake registry.
    home-manager.inputs.nixpkgs.follows = "nixpkgs";
  };

  outputs = {
    nixpkgs,
    home-manager,
    flake-utils,
    ...
  }: 
    flake-utils.lib.eachSystem flake-utils.lib.allSystems (system: {
      # Specify your Home Manager profiles. Home Manager will by
      # default match first `$USER@$(hostname)`. If not found, it
      # will then default to `$USER`.
      packages.homeConfigurations = {
        "marty@Mcfly-MacBook-Pro.local" = home-manager.lib.homeManagerConfiguration {
          pkgs = nixpkgs.legacyPackages.${system};

          # Specify your Home Manager configurations.
          modules = [./host/Mcfly-MacBook-Pro/user/marty/home.nix];

          # Specify extra arguments to pass through your configurations.
        };
      };
    });
}
```

If you want to use symlinks with Nix flakes, see my repo [unvX](https://github.com/pabroux/unvX).

**Comment by joneshf** [comment]
As someone that was just bit by this and spent a non-trivial amount of time trying to come up with a solution, there’s something important that’s missing from these suggestions to move configuration to `homeModules` and re-use it. Namely that it still doesn’t fix the issue of using the same username across multiple systems. There are plenty of valid use cases where you want to have the same username despite the system changing:
- Having more than one computer with the same username (like a linux laptop and a macOS laptop).
- Replacing a computer with a newer model (like updating from Intel macOS to Apple Silicon).
- Ephemeral systems (like multiple Linux boxes that you don’t know or care about the hostname but the architecture can change whether you know about it or not).
- etc.

There are some workarounds you can do sometimes for these cases. But you can’t really easily have two computers with different systems and use the same username for both of them. E.g. If I have both an Intel MacBook Pro and an Apple Silicon MacBook Pro, I can’t easily use the same `homeConfiguration` for these two setups unless I either change the username of one of them, or change the hostname of one of them, make the `flake.nix` way more complicated than it needs to be, or invent some workaround script to switch `home-manager` configurations. None of these workarounds are great, but it doesn’t really seem like they’re actually necessary from a technical standpoint if `home-manager` could change how it resolves `pkgs`.

Since there are workarounds mentioned above, I’m not blocked on this. But this is a surprising situation to run into once you go from a single system to more than one system.

**Comment by ghost** [comment]
For those using `flake-parts`, the above `packages.homeConfigurations` workaround does not work as `perSystem.packages` is typed as `lazyAttrsOf package`. Use `legacyPackages.homeConfigurations` (`lazyAttrsOf raw`) instead.

E.g.:
```nix
{inputs, ...}: {
  perSystem = {pkgs, ...}: {
    legacyPackages.homeConfigurations."<user>" = inputs.home-manager.lib.homeManagerConfiguration {
      inherit pkgs;
      modules = [
        ./home.nix
      ];
    };
  };
}
```

This is ideal if you want to use the same configuration with the same username across arbitrary systems.

---

Alternatively, `home-manager` will also match `<user>@<host>` not just `<user>`. Defining a separate configuration per host may be desirable if you have a known set of machines, especially if their configurations differ.

E.g.:
```nix
{inputs, withSystem, ...}: {
  flake.homeConfigurations = {
    "<user>@<host>" = withSystem "x86_64-linux" ({pkgs, ...}: inputs.home-manager.lib.homeManagerConfiguration {
      inherit pkgs;
      modules = [
        ./common.nix
        ./<host>.nix
      ];
    });
    # "<user>@<host-2>" = ...
    # Etc.
  };
}
```

---

Another option may be to share a modules flake and have a separate, locally defined configuration flake per host.

E.g.:
```nix
# ~/.config/home-manager/flake.nix
{
  inputs = {
    nixpkgs = {
      url = "github:nixos/nixpkgs/nixos-unstable";
    };
    home-manager = {
      url = "github:nix-community/home-manager";
      inputs.nixpkgs.follows = "nixpkgs";
    };
    home = {
      url = "<url>";
      inputs.nixpkgs.follows = "nixpkgs";
      inputs.home-manager.follows = "home-manager";
    };
  };

  outputs = inputs: let
    username = "<user>";
    homeDirectory = "<path>";
    system = "<system>";
  in {
    homeConfigurations.${username} = inputs.home-manager.lib.homeManagerConfiguration {
      pkgs = inputs.nixpkgs.legacyPackages.${system};
      modules = with inputs.home.homeModules; [
        {home = {inherit username homeDirectory;};}
        common
        desktop
        # Etc.
      ];
    };
  };
}
```

This could of course be simplified by outputting a shared lib function to remove the boilerplate from the local host flakes.

E.g.:

```nix
# ~/.config/home-manager/flake.nix
{
  inputs.home.url = "<url>";

  ouputs = inputs:
    inputs.home.lib.mkHome {
      username = "<user>";
      homeDirectory = "<path>";
      system = "<system>";
      modules = with inputs.home.homeModules; [
        common
        desktop
        # Etc.
      ];
    };
}
```

```nix
# <url>/flake.nix
{
  inputs = {
    nixpkgs = {
      url = "github:nixos/nixpkgs/nixos-unstable";
    };
    home-manager = {
      url = "github:nix-community/home-manager";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };

  outputs = inputs: {
    lib.mkHome = {
      username,
      homeDirectory ? "/home/${username}",
      system,
      modules ? [],
    }: {
      homeConfigurations.${username} = inputs.home-manager.lib.homeManagerConfiguration {
        pkgs = inputs.nixpkgs.legacyPackages.${system};
        modules = [{home = {inherit username homeDirectory;};}] ++ modules;
      };
    };

    homeModules = {
      # ...
    };
  };
}
```

**Comment by t3dodson** [comment]
I attempted to add support for multiple systems in a backward compatible way.

I was trying to use flake-parts and ran into this same issue mentioned here. I want one configuration that can work on multiple hosts, and systems without explicitly configuring each of them. 

I’m open to ideas if there is a better way to achieve this end result of decoupling package selection from home-manager configuration. I thought this approach seemed reasonable, if this idea is good I could continue working on it and follow the contribution guide, update docs, etc…

``` 
homeConfigurations.${system}.${username} =
   home-manager.lib.homeManagerConfiguration {
```

```
homeConfigurations.${username} =
  home-manager.lib.homeManagerConfiguration {
```

I was able to switch into my “system” config, we can update the home-manager cli to auto detect this if the base config is not present. The script is obviously not final, but it works for demonstration purposes.

Anyway, I would like to see this issue resolved, and am curious to hear feedback.

https://github.com/nix-community/home-manager/compare/master...t3dodson:home-manager-base:multi-arch

Thanks!

**Comment by nelind3** [comment]
I've been meaning to make an issue just like this and just found out this issue exists so I'll put my two cents here too. Like others have mentioned by far most home configurations act much more like (nixpkgs style) package definitions in that, while a specific instance of a configuration depends on the system its built for, when identifying the config by username the system generally doesn't matter. Therefore I strongly believe `homeConfigurations.${system}.${username}` is a much much more appropriate output path. The reason nixos doesn't do this is that even when identifying a nixos config "by name" it's tied to a specific host with a specific system.

@Shados brings up a good point in that sometimes we genuinely do want the same username to have two different configurations depending on the system. For that usecase I think either `homeConfigurations.${hostname}.${username}` or `homeConfigurations."${username}@${hostname}"` makes the most sense. When a specific host is defined that comes with a specific system too so no need to include that in the path. The former has the upside of aligning with how flakes handle systems. Though *technically* does conflict with the non-host specific path if for whatever strange reason someone really wants their machine to have the hostname `x86_64-linux`.

Backwards compatibility is also a concern, at least for a time after the change is made, if not forever. Therefore my suggestion is to change the CLI to look for the configuration under, in order of priority:
- `homeConfigurations.${current hostname}.${current username}` (or alternatives) it's the most specific. Check that first.
- `homeConfigurations.${current system}.${current username}` the one most people will be using.
- `homeConfigurations."${current username}` backwards compatibility fallback. Can be removed down the road.

This should support every use case in a way that properly supports standalone home manager configurations that are used across multiple hosts that have different systems.
